File: 1580-shuffle-the-array/1580-shuffle-the-array.py
class Solution:
    def shuffle(self, nums: List[int], n: int) -> List[int]:
        for i in range(n):
            nums[i] = nums[i] << 10 | nums[n + i]
        
        end = 2 * n - 1
        for i in range(n - 1, -1, -1):
            x = nums[i] >> 10
            y = nums[i] & 0b1111111111

            nums[end] = y
            nums[end - 1] = x 
            end -= 2
        
        return nums

File: 1580-shuffle-the-array/1580-shuffle-the-array.cpp
class Solution {
public:
    vector<int> shuffle(vector<int>& nums, int n) {
        for (int i = 0; i < n; i++) {
            nums[i] += nums[n + i] << 16;
        }
        // 0 1 2 3 4 5
        for (int i = nums.size() - 1; i >= 0 ; i -= 2) {
            nums[i] = nums[(i - 1) / 2] >> 16;
            nums[i - 1] = nums[(i - 1) / 2] & (0x0000FFFF);   
        }
        return nums;
    }
};